### DISCUSSION TITLE
Report - (3rd Sem.) Week 7 📢

### DISCUSSION THREAD

**Discussion by giomskii** [SE-TINF22B6/DHBWhub#183] [Weekly Report]
Welcome to the seventh report of our progress. We are excited to share the latest news with you! 🥳

### Capturing our architecture significant requirements 🛠️ 
Last week, we discussed about important quality attributes regarding our project by capturing the most essential scenarios. Thus, we created a table regarding many individual events which include the end users or developers and thought about methods which we could use to counter occuring problems. You can either see the full table [here](https://github.com/SE-TINF22B6/DHBWhub/blob/master/docs/ASR.md) or some examples in the following picture:

>![asr-table](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/22a0fb17-b093-48bd-8ae8-2b409ad472a2)

From these quality attributes, we have extracted three important use cases and created an utility tree which you can see below and will also be described in more detail afterwards. We have also updated our [srs](https://github.com/SE-TINF22B6/DHBWhub/blob/master/docs/SOFTWARE_REQUIREMENTS_SPECIFICATION.md#3-nonfunctional-requirements) with the non-functional requirements section, if you are even more interested!

>![utility-tree](https://github.com/SE-TINF22B6/DHBWhub/blob/master/docs/diagrams/UtilityTree.drawio.png?raw=true)

#### Quality attribute 1 - Feedback 💬 
Throughout the whole application, the usabilty of our web application should be simple and easy to understand. One way to achieve this goal is through visual feedback. If a user changes his settings or tries to create new posts and comments, he should get a visual hint, whether his transactions applied successfully or not e.g. a temporary label 'post has been successfully created' or 'Something has gone wrong'.   
The amount of information displayed also needs to be balanced. It should be as little as possible yet enough to help the user if problems occur. One example would be the attempt to sign in onto our website. If the credentials are wrong, we tell the user that the username or the password is incorrect without further hints that could indicate our database architecture to provide more security!

#### Quality attribute 2 - Modularity ♻️ 
This attribute rather appeals to the developers but also includes the end users. Our main design should stay modular in order to allow easier und faster development of new features. On the one hand, it helps the end users to navigate through our application without getting confused. On the other hand, developers can extend the application with new code components without disturbing the already existing system. One way to reach this goal is through an previously established CI/CD pipeline which already covers most areas. Another way is through atomic component architecture. The web design is composed of several small components which are composed of even smaller different components. By reusing already programmed assets, the addition of new features should reduce the development time drastically.

#### Quality attribute 3 - Quick response ⏩ 
If an end user wants to use the features of our website or requests certain components, the response time shouldn't take more than a maximum of 10 seconds (our first estimation, might change later). High latency worsens the mood of the end user. Thus, some structures need to be established to be responsive all the time:
- Giving visual feedback and stopping request, if response time takes too long.
- Using intermediate data layers where important data is already stored and only waits for a request.
- Already storing frequently requested data in a database view which can be delivered immediately when needed.
- Asynchronous loading in the background if the user doesn't need the feature at the moment but might use it later on.

### Finished calendar mockup 📆 
We have finished the mockup to our calendar which contains events that could be interesting for university students. We want to keep the calendar static at the moment which means that every student can see the same events.  
Right below is the monthly view of the calendar:
![Calendar Month- PC View](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/b2c543a0-c308-494e-a016-ad54dfce7737)

This is the weekly view of the calendar:

![Calendar Week - PC View](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/68deb5d9-fa99-4ee0-9734-59b6ac19eaae)

### Post with event mockup 📅 
We have created a new version of our post mockup which displays an event post. A map with the event location is also included:
![Calendar Event- PC View](https://github.com/SE-TINF22B6/DHBWhub/assets/111356450/d99aff7c-7102-42ac-b53a-dffc964dfa9c)

### Implementation of login/sign-up window 🪟 
The first step towards the actual sign-in/up process is ready! The windows are designed just like the respective mock-ups, the functioniality however is not implemented yet. You can see them in the following pictures:
>![login-process](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/4666085f-7b10-4d2b-99a0-3fcae413e567)
>![registration-process](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/b523570f-e8da-4496-a8d8-d56893c0b8e9)

### Implementation of contact form 📞 
We have finished the actual implementation of our contact form. You have to enter your e-mail, name and message to send it to the site administrators!
>![grafik](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/1a60c334-01bf-404e-8545-4f5faca41f12)
 

### Sprint 4: What have we achieved ❓ What's to come❔
✅ Finish Calendar Mockup in Figma
✅ Implement Login Window
✅ Implement Sign Up Window
✅ Implement contact page
🚧 Discuss search opportunities and create mockup
🚧 Implement post menu
🚧 Implement first frontend endpoints
🚧 Implement e-mail verification
🚧 Setup postman for whole team
🚧 Create endpoints for search
🚧 Create endpoints for the sort options of posts
🚧 Removing Herobrine

Until next week 💯

Your DHBWhub Team 🚀

**Comment by ins305** [comment]
Hi Team DHBWhub,

You have a very good collection of different quality attributes. Also, your top 3 scenarios are a good choice for your project as they cover important topics. 
Overall, your descriptions of the quality attributes are very clear and detailed. Good job!

I can see that you also put a lot of effort into your really nice design for your project. 

Keep up the good work!

**Comment by dolerich-hirnfiedler** [comment]
Dear DHBWhub Team,

Impressive progress report! 🌟 Your detailed analysis and strategic approach to quality attributes showcase a deep understanding of project intricacies.
 The utility tree visualizes key use cases effectively, emphasizing user-friendly feedback, modularity, and quick response times.
 Your commitment to simplicity, modular design, and efficient development through CI/CD and atomic components is commendable. The proposed strategies for quick response times reveal a keen awareness of user satisfaction.
 The finished calendar mockup adds a tangible highlight to your achievements. Kudos to the team for a thoughtful and impactful update! 🚀👏

**Comment by LeonmAndreesen** [comment]
Hi, 
your weekly update is stunning. For sure, the best report so far. It is so detailed. Moreover the amount of tasks, that you achieved is superb. 
Your mock-ups are really great to look at. Nicely chosen colors enhance the visual perception of the first impression. 
Even though you did not implement functions on them, you are on a good way.
